src/main/java/tech/stackable/gis/hbase/coprocessor/KNNEndpoint.java

```diff
@@ -0,0 +1,135 @@
+import org.apache.hadoop.hbase.coprocessor.CoprocessorService;
+import org.apache.hadoop.hbase.coprocessor.RegionCoprocessor;
+import org.apache.hadoop.hbase.coprocessor.RegionCoprocessorEnvironment;
+import org.apache.hadoop.hbase.regionserver.InternalScanner;
+import org.apache.hadoop.hbase.shaded.protobuf.ResponseConverter;
+import org.apache.hadoop.hbase.util.Bytes;
+import org.apache.hbase.thirdparty.com.google.common.collect.MinMaxPriorityQueue;
+import tech.stackable.gis.hbase.generated.KNN;
+
+import java.io.IOException;
+import java.util.ArrayList;
+import java.util.List;
+
+public class KNNEndpoint extends KNN.KNNService implements RegionCoprocessor, CoprocessorService {
+    static final Log LOG = LogFactory.getLog(KNNEndpoint.class);
+
+    private RegionCoprocessorEnvironment env;
+
+    @Override
+    public Service getService() {
+        return this;
+    }
+
+    @Override
+    public void start(CoprocessorEnvironment env) throws IOException {
+        if (env instanceof RegionCoprocessorEnvironment) {
+            this.env = (RegionCoprocessorEnvironment) env;
+        } else {
+            throw new CoprocessorException("Must be loaded on a table region!");
+        }
+    }
+
+    @Override
+    public void stop(CoprocessorEnvironment env) throws IOException {
+        // do nothing
+    }
+
+    @Override
+    public void getKNN(RpcController controller, KNN.KNNRequest request, RpcCallback<KNN.KNNResponse> done) {
+
+        final byte[] family = request.getFamily().toByteArray();
+        final byte[] latCol = request.getLatCol().toByteArray();
+        final byte[] lonCol = request.getLonCol().toByteArray();
+        final int count = request.getCount();
+
+        Scan scan = new Scan();
+        scan.addFamily(family);
+        scan.addColumn(family, latCol);
+        scan.addColumn(family, lonCol);
+
+        KNN.KNNResponse response = null;
+        InternalScanner scanner = null;
+        final var distComp = new DistComp(request.getLat(), request.getLon());
+        final MinMaxPriorityQueue<Neighbor> knns = MinMaxPriorityQueue.orderedBy(distComp).maximumSize(count).create();
+        try {
+            scanner = env.getRegion().getScanner(scan);
+            List<Cell> results = new ArrayList<>();
+            boolean hasMore;
+
+            do {
+                var lon = Double.NaN;
+                var lat = Double.NaN;
+                hasMore = scanner.next(results);
+                for (Cell cell : results) {
+                    if (CellUtil.matchingColumn(cell, family, lonCol))
+                        lon = parseCoordinate(cell);
+                    if (CellUtil.matchingColumn(cell, family, latCol))
+                        lat = parseCoordinate(cell);
+
+                    if (!Double.isNaN(lat) && !Double.isNaN(lon)) {
+                        knns.add(new Neighbor(Bytes.toString(cell.getRowArray(), cell.getRowOffset(), cell.getRowLength()), lat, lon));
```

Comment:
See above: We want to avoid creating the Neighbor object for everything even those things we immediately throw away again.

So I think keeping a PriorityQueue plus remembering the largest distance and then doing everything manually might be better